File: apps/ios/Tests/CrossyUITests/ReactionFanModelTests.swift
// The fan's grammar, exhaustively (written before any gesture code, per the wave
// plan): hold-slide-release fires on release over an emoji and cancels anywhere else;
// releasing on the button is the tap fallback into a standing fan; a standing fan
// fires on tap, yields to a tap away, toggles closed from its own button, and folds
// after ~3 s idle. Firing always dismisses (owner ruling from the web review). The
// capsule layout's hit math is pinned beside it so render and hit test cannot drift.

import CrossyUI
import XCTest

final class ReactionFanModelTests: XCTestCase {
    // MARK: - Hold-slide-release

    func test_startsClosed() {
        let fan = ReactionFanModel()
        XCTAssertEqual(fan.phase, .closed)
        XCTAssertFalse(fan.isOpen)
    }

    func test_holdOpensImmediately() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        XCTAssertEqual(fan.phase, .heldOpen)
    }

    func test_holdMovedHighlightsTheSlotUnderTheFinger() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        fan.holdMoved(over: 2)
        XCTAssertEqual(fan.highlighted, 2)
        fan.holdMoved(over: nil)
        XCTAssertNil(fan.highlighted)
        fan.holdMoved(over: 99)
        XCTAssertNil(fan.highlighted, "an out-of-range slot never highlights")
    }

    func test_holdMovedIsIgnoredWhileClosed() {
        var fan = ReactionFanModel()
        fan.holdMoved(over: 1)
        XCTAssertNil(fan.highlighted)
        XCTAssertEqual(fan.phase, .closed)
    }

    func test_releaseOverAnEmojiFiresAndDismisses() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        fan.holdMoved(over: 0)
        let effect = fan.holdEnded(over: 0, onButton: false, at: 10)
        XCTAssertEqual(effect, .fire("🔥"), "slot 1 of the D25 defaults")
        XCTAssertEqual(fan.phase, .closed, "firing always dismisses the fan")
        XCTAssertNil(fan.highlighted)
    }

    func test_releaseElsewhereCancels() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        fan.holdMoved(over: 3)
        let effect = fan.holdEnded(over: nil, onButton: false, at: 10)
        XCTAssertEqual(effect, .none)
        XCTAssertEqual(fan.phase, .closed)
    }

    func test_releaseWithAnInvalidSlotCancels() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        let effect = fan.holdEnded(over: 99, onButton: false, at: 10)
        XCTAssertEqual(effect, .none)
        XCTAssertEqual(fan.phase, .closed)
    }

    func test_holdEndedWhileClosedIsANoOp() {
        var fan = ReactionFanModel()
        XCTAssertEqual(fan.holdEnded(over: 0, onButton: false, at: 10), .none)
        XCTAssertEqual(fan.phase, .closed)
    }

    // MARK: - The tap fallback (release on the button)

    func test_releaseOnTheButtonOpensTheStandingFan() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        let effect = fan.holdEnded(over: nil, onButton: true, at: 10)
        XCTAssertEqual(effect, .none)
        XCTAssertEqual(fan.phase, .tapOpen)
        XCTAssertEqual(fan.openedAt, 10)
    }

    func test_tapOnTheButtonTogglesAStandingFanClosed() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        _ = fan.holdEnded(over: nil, onButton: true, at: 10)
        // The second tap: hold begins on a standing fan, releases on the button.
        fan.holdBegan()
        let effect = fan.holdEnded(over: nil, onButton: true, at: 11)
        XCTAssertEqual(effect, .none)
        XCTAssertEqual(fan.phase, .closed, "the button toggles, never re-opens")
    }

    func test_holdFromAStandingFanCanStillSlideAndFire() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        _ = fan.holdEnded(over: nil, onButton: true, at: 10)
        fan.holdBegan()
        fan.holdMoved(over: 4)
        let effect = fan.holdEnded(over: 4, onButton: false, at: 12)
        XCTAssertEqual(effect, .fire("😭"), "slot 5 of the D25 defaults")
        XCTAssertEqual(fan.phase, .closed)
    }

    // MARK: - Configurable slots (D25: the fan wears the personal set)

    func test_defaultFanWearsTheD25DefaultFive_PROTOCOL9() {
        XCTAssertEqual(ReactionFanModel().emojis, ReactionPolicy.defaultSet)
        XCTAssertEqual(ReactionFanModel().emojis, ["🔥", "🤔", "🐐", "💀", "😭"])
    }

    func test_customSlotsFireTheCustomGrapheme_D25() {
        // A personal set in slot order: every fire path returns the slot's OWN
        // grapheme, so the fan is entirely set-agnostic (skin tones included:
        // distinctness and identity are the exact strings, PROTOCOL.md §12).
        let personal = ["🦆", "👍🏽", "❤️‍🔥", "🇨🇦", "🫶"]
        var fan = ReactionFanModel(emojis: personal)
        fan.holdBegan()
        fan.holdMoved(over: 2)
        XCTAssertEqual(fan.holdEnded(over: 2, onButton: false, at: 10), .fire("❤️‍🔥"))

        fan.holdBegan()
        _ = fan.holdEnded(over: nil, onButton: true, at: 12)
        XCTAssertEqual(fan.tapEmoji(at: 3), .fire("🇨🇦"))
    }

    // MARK: - The standing fan

    func test_tapOnAnEmojiFiresAndDismisses() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        _ = fan.holdEnded(over: nil, onButton: true, at: 10)
        let effect = fan.tapEmoji(at: 1)
        XCTAssertEqual(effect, .fire("🤔"), "slot 2 holds 🤔 in the defaults, old and new")
        XCTAssertEqual(fan.phase, .closed, "firing always dismisses the fan")
    }

    func test_tapEmojiWhileClosedIsANoOp() {
        var fan = ReactionFanModel()
        XCTAssertEqual(fan.tapEmoji(at: 1), .none)
        XCTAssertEqual(fan.phase, .closed)
    }

    func test_tapAwayClosesTheStandingFanOnly() {
        var fan = ReactionFanModel()
        fan.tapAway()
        XCTAssertEqual(fan.phase, .closed)
        fan.holdBegan()
        _ = fan.holdEnded(over: nil, onButton: true, at: 10)
        fan.tapAway()
        XCTAssertEqual(fan.phase, .closed)
    }

    func test_idleTimeoutClosesAtThreeSeconds() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        _ = fan.holdEnded(over: nil, onButton: true, at: 10)
        fan.idleExpired(at: 10 + ReactionFanModel.tapOpenIdleSeconds - 0.1)
        XCTAssertEqual(fan.phase, .tapOpen, "not idle yet")
        fan.idleExpired(at: 10 + ReactionFanModel.tapOpenIdleSeconds)
        XCTAssertEqual(fan.phase, .closed)
    }

    func test_aStaleIdleTimerCannotCloseANewerOpening() {
        var fan = ReactionFanModel()
        fan.holdBegan()
        _ = fan.holdEnded(over: nil, onButton: true, at: 20)
        // A timer scheduled against an earlier opening fires with an instant that
        // precedes this fan's deadline: validated against openedAt, it is a no-op.
        fan.idleExpired(at: 21)
        XCTAssertEqual(fan.phase, .tapOpen)
    }

    // MARK: - The capsule layout (render and hit test share one geometry)

    func test_slotCentersFallInsideTheirOwnSlots() {
        let count = ReactionPolicy.defaultSet.count
        for index in 0..<count {
            let x = ReactionFanLayout.slotCenterX(index: index, count: count)
            XCTAssertEqual(
                ReactionFanLayout.slot(atX: x, y: ReactionFanLayout.height / 2, count: count),
                index)
        }
    }

    func test_slotHitsClampWithinTheSlackAndDieBeyondIt() {
        let count = 5
        let width = ReactionFanLayout.width(count: count)
        // Just past the leading edge, inside the slack: the first slot.
        XCTAssertEqual(ReactionFanLayout.slot(atX: -5, y: 22, count: count), 0)
        // Just past the trailing edge, inside the slack: the last slot.
        XCTAssertEqual(ReactionFanLayout.slot(atX: width + 5, y: 22, count: count), count - 1)
        // Beyond the slack in x or y: off the row.
        XCTAssertNil(
            ReactionFanLayout.slot(
                atX: -ReactionFanLayout.holdSlack - 1, y: 22, count: count))
        XCTAssertNil(
            ReactionFanLayout.slot(
                atX: 22, y: ReactionFanLayout.height + ReactionFanLayout.holdSlack + 1,
                count: count))
    }

    func test_slotSizeMeetsTheTouchFloor() {
        // 44 pt minimum tap target (the design-engineering floor); the hold-slide
        // slack widens it further for the occluding thumb.
        XCTAssertGreaterThanOrEqual(ReactionFanLayout.slotSize, 44)
    }
}

File: apps/android/ui/src/test/kotlin/crossy/ui/ReactionFanModelTests.kt
// The fan's grammar, exhaustively (written before any gesture code). Twin of apps/ios
// ReactionFanModelTests.swift: hold-slide-release fires on release over an emoji and cancels
// anywhere else; releasing on the button is the tap fallback into a standing fan; a standing fan
// fires on tap, yields to a tap away, toggles closed from its own button, and folds after ~3 s idle.
// Firing always dismisses (owner ruling from the web review). The capsule layout's hit math is
// pinned beside it so render and hit test cannot drift. The Kotlin model is a pure value type, so
// each transition returns the next model (or a Step carrying the effect).

package crossy.ui

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Test

class ReactionFanModelTests {
    // --- Hold-slide-release ---

    @Test
    fun startsClosed() {
        val fan = ReactionFanModel()
        assertEquals(ReactionFanModel.Phase.CLOSED, fan.phase)
        assertFalse(fan.isOpen)
    }

    @Test
    fun holdOpensImmediately() {
        val fan = ReactionFanModel().holdBegan()
        assertEquals(ReactionFanModel.Phase.HELD_OPEN, fan.phase)
    }

    @Test
    fun holdMovedHighlightsTheSlotUnderTheFinger() {
        var fan = ReactionFanModel().holdBegan()
        fan = fan.holdMoved(over = 2)
        assertEquals(2, fan.highlighted)
        fan = fan.holdMoved(over = null)
        assertNull(fan.highlighted)
        fan = fan.holdMoved(over = 99)
        assertNull(fan.highlighted, "an out-of-range slot never highlights")
    }

    @Test
    fun holdMovedIsIgnoredWhileClosed() {
        val fan = ReactionFanModel().holdMoved(over = 1)
        assertNull(fan.highlighted)
        assertEquals(ReactionFanModel.Phase.CLOSED, fan.phase)
    }

    @Test
    fun releaseOverAnEmojiFiresAndDismisses() {
        var fan = ReactionFanModel().holdBegan()
        fan = fan.holdMoved(over = 0)
        val step = fan.holdEnded(over = 0, onButton = false, now = 10.0)
        assertEquals(ReactionFanModel.Effect.Fire("🔥"), step.effect, "slot 1 of the D25 defaults")
        assertEquals(ReactionFanModel.Phase.CLOSED, step.model.phase, "firing always dismisses the fan")
        assertNull(step.model.highlighted)
    }

    @Test
    fun releaseElsewhereCancels() {
        var fan = ReactionFanModel().holdBegan()
        fan = fan.holdMoved(over = 3)
        val step = fan.holdEnded(over = null, onButton = false, now = 10.0)
        assertEquals(ReactionFanModel.Effect.None, step.effect)
        assertEquals(ReactionFanModel.Phase.CLOSED, step.model.phase)
    }

    @Test
    fun releaseWithAnInvalidSlotCancels() {
        val fan = ReactionFanModel().holdBegan()
        val step = fan.holdEnded(over = 99, onButton = false, now = 10.0)
        assertEquals(ReactionFanModel.Effect.None, step.effect)
        assertEquals(ReactionFanModel.Phase.CLOSED, step.model.phase)
    }

    @Test
    fun holdEndedWhileClosedIsANoOp() {
        val fan = ReactionFanModel()
        val step = fan.holdEnded(over = 0, onButton = false, now = 10.0)
        assertEquals(ReactionFanModel.Effect.None, step.effect)
        assertEquals(ReactionFanModel.Phase.CLOSED, step.model.phase)
    }

    // --- The tap fallback (release on the button) ---

    @Test
    fun releaseOnTheButtonOpensTheStandingFan() {
        val fan = ReactionFanModel().holdBegan()
        val step = fan.holdEnded(over = null, onButton = true, now = 10.0)
        assertEquals(ReactionFanModel.Effect.None, step.effect)
        assertEquals(ReactionFanModel.Phase.TAP_OPEN, step.model.phase)
        assertEquals(10.0, step.model.openedAt)
    }

    @Test
    fun tapOnTheButtonTogglesAStandingFanClosed() {
        var fan = ReactionFanModel().holdBegan()
        fan = fan.holdEnded(over = null, onButton = true, now = 10.0).model
        // The second tap: hold begins on a standing fan, releases on the button.
        fan = fan.holdBegan()
        val step = fan.holdEnded(over = null, onButton = true, now = 11.0)
        assertEquals(ReactionFanModel.Effect.None, step.effect)
        assertEquals(ReactionFanModel.Phase.CLOSED, step.model.phase, "the button toggles, never re-opens")
    }

    @Test
    fun holdFromAStandingFanCanStillSlideAndFire() {
        var fan = ReactionFanModel().holdBegan()
        fan = fan.holdEnded(over = null, onButton = true, now = 10.0).model
        fan = fan.holdBegan()
        fan = fan.holdMoved(over = 4)
        val step = fan.holdEnded(over = 4, onButton = false, now = 12.0)
        assertEquals(ReactionFanModel.Effect.Fire("😭"), step.effect, "slot 5 of the D25 defaults")
        assertEquals(ReactionFanModel.Phase.CLOSED, step.model.phase)
    }

    // --- Configurable slots (D25: the fan wears the personal set) ---

    @Test
    fun defaultFanWearsTheD25DefaultFive_PROTOCOL9() {
        assertEquals(ReactionPolicy.defaultSet, ReactionFanModel().emojis)
        assertEquals(listOf("🔥", "🤔", "🐐", "💀", "😭"), ReactionFanModel().emojis)
    }

    @Test
    fun customSlotsFireTheCustomGrapheme_D25() {
        // A personal set in slot order: every fire path returns the slot's OWN grapheme, so the fan
        // is entirely set-agnostic (skin tones included: distinctness and identity are the exact
        // strings, PROTOCOL.md §12).
        val personal = listOf("🦆", "👍🏽", "❤️‍🔥", "🇨🇦", "🫶")
        var fan = ReactionFanModel(emojis = personal).holdBegan()
        fan = fan.holdMoved(over = 2)
        assertEquals(ReactionFanModel.Effect.Fire("❤️‍🔥"), fan.holdEnded(over = 2, onButton = false, now = 10.0).effect)

        fan = ReactionFanModel(emojis = personal).holdBegan()
        fan = fan.holdEnded(over = null, onButton = true, now = 12.0).model
        assertEquals(ReactionFanModel.Effect.Fire("🇨🇦"), fan.tapEmoji(at = 3).effect)
    }

    // --- The standing fan ---

    @Test
    fun tapOnAnEmojiFiresAndDismisses() {
        var fan = ReactionFanModel().holdBegan()
        fan = fan.holdEnded(over = null, onButton = true, now = 10.0).model
        val step = fan.tapEmoji(at = 1)
        assertEquals(ReactionFanModel.Effect.Fire("🤔"), step.effect, "slot 2 holds 🤔 in the defaults")
        assertEquals(ReactionFanModel.Phase.CLOSED, step.model.phase, "firing always dismisses the fan")
    }

    @Test
    fun tapEmojiWhileClosedIsANoOp() {
        val fan = ReactionFanModel()
        val step = fan.tapEmoji(at = 1)
        assertEquals(ReactionFanModel.Effect.None, step.effect)
        assertEquals(ReactionFanModel.Phase.CLOSED, step.model.phase)
    }

    @Test
    fun tapAwayClosesTheStandingFanOnly() {
        var fan = ReactionFanModel().tapAway()
        assertEquals(ReactionFanModel.Phase.CLOSED, fan.phase)
        fan = ReactionFanModel().holdBegan()
        fan = fan.holdEnded(over = null, onButton = true, now = 10.0).model
        fan = fan.tapAway()
        assertEquals(ReactionFanModel.Phase.CLOSED, fan.phase)
    }

    @Test
    fun idleTimeoutClosesAtThreeSeconds() {
        var fan = ReactionFanModel().holdBegan()
        fan = fan.holdEnded(over = null, onButton = true, now = 10.0).model
        fan = fan.idleExpired(now = 10.0 + ReactionFanModel.TAP_OPEN_IDLE_SECONDS - 0.1)
        assertEquals(ReactionFanModel.Phase.TAP_OPEN, fan.phase, "not idle yet")
        fan = fan.idleExpired(now = 10.0 + ReactionFanModel.TAP_OPEN_IDLE_SECONDS)
        assertEquals(ReactionFanModel.Phase.CLOSED, fan.phase)
    }

    @Test
    fun aStaleIdleTimerCannotCloseANewerOpening() {
        var fan = ReactionFanModel().holdBegan()
        fan = fan.holdEnded(over = null, onButton = true, now = 20.0).model
        // A timer scheduled against an earlier opening fires with an instant that precedes this
        // fan's deadline: validated against openedAt, it is a no-op.
        fan = fan.idleExpired(now = 21.0)
        assertEquals(ReactionFanModel.Phase.TAP_OPEN, fan.phase)
    }

    // --- The capsule layout (render and hit test share one geometry) ---

    @Test
    fun slotCentersFallInsideTheirOwnSlots() {
        val count = ReactionPolicy.defaultSet.size
        for (index in 0 until count) {
            val x = ReactionFanLayout.slotCenterX(index, count)
            assertEquals(index, ReactionFanLayout.slot(atX = x, y = ReactionFanLayout.height / 2, count = count))
        }
    }

    @Test
    fun slotHitsClampWithinTheSlackAndDieBeyondIt() {
        val count = 5
        val width = ReactionFanLayout.width(count)
        assertEquals(0, ReactionFanLayout.slot(atX = -5.0, y = 22.0, count = count))
        assertEquals(count - 1, ReactionFanLayout.slot(atX = width + 5, y = 22.0, count = count))
        assertNull(ReactionFanLayout.slot(atX = -ReactionFanLayout.HOLD_SLACK - 1, y = 22.0, count = count))
        assertNull(
            ReactionFanLayout.slot(
                atX = 22.0,
                y = ReactionFanLayout.height + ReactionFanLayout.HOLD_SLACK + 1,
                count = count,
            ),
        )
    }

    @Test
    fun slotSizeMeetsTheTouchFloor() {
        // 44 dp minimum tap target (the design-engineering floor); the hold-slide slack widens it.
        assertEquals(true, ReactionFanLayout.SLOT_SIZE >= 44.0)
    }
}
